Add likes feature

ellaml opened the pull request:
- When a user is logged in it's liked reviews will be marked as liked
- Clicking on the like button when logged in will toggle the like and update the review object as well
- Split to different commits: 
1) adding the data model and tests for it
2) adding displaying likes feature
3) adding updating likes feature

![image](https://user-images.githubusercontent.com/73358484/118841608-2302e500-b8d1-11eb-946c-df1d424c1079.png)

fix: #115

ellaml committed:
Added User_Likes model with test data and tests

- Added User_Likes model, a table with user_id and review_id for user
  likes
- Added test data for the new table
- Added tests for the test data and the model methods

ellaml committed:
Added display of like button for liked reviews

- Reviews which the logged in user liked now displayed with glowing
  button and updated text in pages Reviews, and Single Course

ellaml committed:
Added likes functionality

- Clicking on the like button when a user is logged in will update the
  review in db and rerender the same page
- Like button is disabled for unauthorized user

ellaml committed:
resolved conflicts with main

ellaml committed:
fixed conflicts that failed tests

oriAdler commented on homepage/models.py:
```diff
@@ -292,3 +292,42 @@ def landing_page_feed(cls):
+        self.save(update_fields=['likes_num'])
+
+
+class User_Likes(models.Model):
```
```suggestion
class UserLikes(models.Model):

```

ellaml replied:
@oriAdler changed

oriAdler commented on homepage/views.py:
```diff
@@ -47,7 +48,11 @@ def courses(request):
 def reviews(request):
     reviews = Review.main_feed()
-    return render(request, 'homepage/reviews/reviews.html', {'reviews': reviews})
+    if(request.user.is_anonymous):
```
```suggestion
    liked_reviews = [] if request.user.is_anonymous else User_Likes.get_liked_reviews_by_user(request.user)
```
WDYT?

ellaml replied:
@oriAdler left it as it was, I think it's more readable

oriAdler commented on homepage/views.py:
```diff
@@ -77,7 +82,16 @@ def course(request, id):
         course = Course.objects.get(pk=id)
         reviews = Review.objects.filter(course=id).order_by('-likes_num')
-        return render(request, 'homepage/courses/course.html', {'id': id, 'course': course, 'reviews': reviews})
+        if(request.user.is_anonymous):
```
Same ^^

VeraTs replied:
this would be too long (123 chars) and she would have to cut it midway, feels more readable like this

ellaml replied:
@oriAdler same

oriAdler reviewed (commented):
Looks great! 
I've left some comments

VeraTs commented on homepage/models.py:
```diff
@@ -292,3 +292,42 @@ def landing_page_feed(cls):
         return True if Review.objects.filter(user=user_id, course=course_id) else False
+
+    def add_one_like(self):
+        self.likes_num = self.likes_num + 1
```
I think add_like and remove_like sound nicer - but this is just a suggestion

ellaml replied:
@VeraTs changed

VeraTs reviewed (commented):
Looks Great!
Left some comments, but I didn't see any tests for the view?

ellaml committed:
Changed model name and added view tests

- Changed from User_Likes to UserLikes
- Added test file for the view with 9 tests
- Small fixes

ellaml committed:
Fixed flake8 errors

oriAdler reviewed (approved):
LGTM

LeonRabinovich commented:
LGTM

kobihk commented on homepage/migrations/0012_user_likes_test_data.py:
```diff
@@ -0,0 +1,37 @@
+            for user, review in likes_test_data:
+                UserLikes(
+                    user_id=user,
+                    review_id=review).save()
```
```suggestion
                UserLikes(user_id=user, review_id=review).save()

```

kobihk commented on homepage/models.py:
```diff
@@ -292,3 +292,42 @@ def landing_page_feed(cls):
         return True if Review.objects.filter(user=user_id, course=course_id) else False
+
+    def add_like(self):
+        self.likes_num = self.likes_num + 1
```
```suggestion
        self.likes_num +=  1

```

kobihk commented on homepage/models.py:
```diff
@@ -292,3 +292,42 @@ def landing_page_feed(cls):
+        self.save(update_fields=['likes_num'])
+
+    def remove_like(self):
+        self.likes_num = self.likes_num - 1
```
```suggestion
        self.likes_num -= 1

```

kobihk commented on homepage/tests/user_likes_model_tests.py:
```diff
@@ -0,0 +1,69 @@
+def user_likes_list():
+    return list(UserLikes.objects.values_list(
+        'user_id',
+        'review_id'))
```
```suggestion
    return list(UserLikes.objects.values_list('user_id', 'review_id'))

```

kobihk commented on homepage/tests/user_likes_view_tests.py:
```diff
@@ -0,0 +1,99 @@
+    return User.objects.get(pk=2), Review.objects.get(pk=5)
+
+
+# check db when using view from reviews page
+@pytest.mark.django_db
+def test_view_toggle_like_from_reviews(client, db_data):
+    likes_num_before = db_data[1].likes_num
+    client.get(f'/like/{db_data[0].id}/{db_data[1].id}/', HTTP_REFERER='/reviews/')
+    assert UserLikes.objects.filter(user_id=db_data[0], review_id=db_data[1]).exists()
+    review = Review.objects.get(pk=5)  # pull updated
+    assert review.likes_num == likes_num_before + 1
+
+
+# check db when using view from single course page
+@pytest.mark.django_db
+def test_view_toggle_like_from_course(client, db_data):
+    likes_num_before = db_data[1].likes_num
+    client.get(f'/like/{db_data[0].id}/{db_data[1].id}/', HTTP_REFERER='/course/10111/')
+    assert UserLikes.objects.filter(user_id=db_data[0], review_id=db_data[1]).exists()
+    review = Review.objects.get(pk=5)  # pull updated
+    assert review.likes_num == likes_num_before + 1
```
AFAICS these tests are the same except the HTTP_REFERER 
maybe you can do it without repetition with @pytest.mark.parametrize

kobihk commented on homepage/tests/user_likes_view_tests.py:
```diff
@@ -0,0 +1,99 @@
+    assert review.likes_num == likes_num_before + 1
+
+
+# check db when using invalid user id from valid reviews page
+@pytest.mark.django_db
+def test_view_toggle_like_invalid_user(client, db_data):
+    likes_num_before = db_data[1].likes_num
+    client.get('/like/10/3/', HTTP_REFERER='/reviews/')  # valid referrer
+    assert not UserLikes.objects.filter(user_id=db_data[0], review_id=db_data[1]).exists()
+    review = db_data[1]
+    assert review.likes_num == likes_num_before
+
+
+# check db when using invalid review id from valid reviews page
+@pytest.mark.django_db
+def test_view_toggle_like_invalid_review(client, db_data):
+    likes_num_before = db_data[1].likes_num
+    client.get('/like/3/30/', HTTP_REFERER='/reviews/')  # valid referrer
+    assert not UserLikes.objects.filter(user_id=db_data[0], review_id=db_data[1]).exists()
+    review = db_data[1]
+    assert review.likes_num == likes_num_before
```
same ^^

kobihk commented on homepage/tests/user_likes_view_tests.py:
```diff
@@ -0,0 +1,99 @@
+    assert review.likes_num == likes_num_before
+
+
+# check that it stays on the same page - reviews
+@pytest.mark.django_db
+def test_rendering_toggle_like_from_reviews(client):
+    response = client.get('/like/2/5/', HTTP_REFERER='/reviews/')
+    assert response.status_code == 302
+    response = client.get(response.url)
+    assert response.status_code == 200
+    assertTemplateUsed(response, 'homepage/reviews/reviews.html')
+
+
+# check that it stays on the same page - single course
+@pytest.mark.django_db
+def test_rendering_toggle_like_from_course(client):
+    response = client.get('/like/2/5/', HTTP_REFERER='/course/10111/')
+    assert response.status_code == 302
+    response = client.get(response.url)
+    assert response.status_code == 200
+    assertTemplateUsed(response, 'homepage/courses/course.html')
```
same ^^

kobihk commented on homepage/tests/user_likes_view_tests.py:
```diff
@@ -0,0 +1,99 @@
+    assert response.status_code == 302
+    response = client.get(response.url)
+    assert response.status_code == 200
+    assertTemplateUsed(response, 'homepage/landing/landing.html')
```
same ^^

kobihk commented on homepage/views.py:
```diff
@@ -47,7 +48,11 @@ def courses(request):
 def reviews(request):
     reviews = Review.main_feed()
-    return render(request, 'homepage/reviews/reviews.html', {'reviews': reviews})
+    if (request.user.is_anonymous):
+        liked_reviews = []
+    else:
+        liked_reviews = UserLikes.get_liked_reviews_by_user(request.user)
```
because I saw you didn't like the same comment @oriAdler added
```suggestion
        liked_reviews = [] if request.user.is_anonymous else UserLikes.get_liked_reviews_by_user(request.user)
```

I added one more option:
```suggestion
        liked_reviews = []
        if not request.user.is_anonymous:
             liked_reviews = UserLikes.get_liked_reviews_by_user(request.user)
```
and if you don't like it either so np

kobihk commented on homepage/views.py:
```diff
@@ -77,7 +82,16 @@ def course(request, id):
         reviews = Review.objects.filter(course=id).order_by('-likes_num')
-        return render(request, 'homepage/courses/course.html', {'id': id, 'course': course, 'reviews': reviews})
+        if(request.user.is_anonymous):
+            liked_reviews = []
+        else:
+            liked_reviews = UserLikes.get_liked_reviews_by_user_for_course(request.user, course)
```
same ^^

ellaml committed:
added parmetrize in view tests

ellaml committed:
resolved conflicts with main

kobihk commented on homepage/static/css/add_review.css:
```diff
@@ -0,0 +1,44 @@
+.control-group {
+	margin: 10px 0 20px 0;
+
+}
```
please check if you missing a new line at the end of the file

ellaml committed:
fix missing line and the end of file

kobihk commented:
@ellaml please solve the conflicts

ellaml committed:
resolved conflicts with main

ellaml commented:
@kobihk resolved
